### PR TITLE
Transaction support

### PR DESCRIPTION
**Pull request by zachmu**
Only START TRANSACTION and COMMIT are supported.

Needs SAVEPOINT, ROLLBACK.

Also needs tests.

### CODE REVIEW

File: go/store/datas/database_common.go [hunk 1]
```diff
@@ -521,6 +521,103 @@ func (db *database) doTag(ctx context.Context, datasetID string, tag types.Struc
...
+
+			return db.doUpdateWorkingSet(ctx, ds.ID(), workspace, prevHash)
+		},
+	)
+}
+
+// doUpdateWorkingSet manages concurrent access the single logical piece of mutable state: the current Root. It uses
+// the same optimistic locking write algorithm as doCommit (see above). Unlike doCommit and other methods in this file,
+// an error is returned if the current value of the ref being written has changed.
+// Workspace updates are serialized, but all other
```
Comment (andrew-wm-arthur) [inline go/store/datas/database_common.go:542]: unfinished comment?

File: go/libraries/doltcore/sqle/transactions.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+		if rootsEqual(root, tx.startRoot) {
+			// ff merge
+			err = tx.db.UpdateWorkingSet(ctx, tx.workingSet, newRoot, hash)
+			if err == datas.ErrOptimisticLockFailed {
+				continue
+			}
+		}
+
+		mergedRoot, _, err := merge.MergeRoots(ctx, root, newRoot, tx.startRoot)
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/transactions.go:89]: how are we handling merge conflicts here?

**Review by andrew-wm-arthur — COMMENTED** [review]
General approach seems reasonable, couple comments

**Comment by andrew-wm-arthur** [comment]
#benchmark

**Comment by andrew-wm-arthur** [comment]
#benchmark

File: go/libraries/doltcore/doltdb/doltdb.go [hunk 3]
```diff
@@ -731,9 +753,16 @@ func (ddb *DoltDB) GetWorkspaces(ctx context.Context) ([]ref.DoltRef, error) {
 	return ddb.GetRefsOfType(ctx, workspacesRefFilter)
 }
 
-// GetRefs returns a list of all refs in the database.
-func (ddb *DoltDB) GetRefs(ctx context.Context) ([]ref.DoltRef, error) {
-	return ddb.GetRefsOfType(ctx, ref.RefTypes)
+var workingSetsRefFilter = map[ref.RefType]struct{}{ref.WorkingSetRefType: {}}
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/doltdb/doltdb.go:756]: Dead now

File: go/libraries/doltcore/doltdb/doltdb.go [hunk 4]
```diff
@@ -731,9 +753,16 @@ func (ddb *DoltDB) GetWorkspaces(ctx context.Context) ([]ref.DoltRef, error) {
...
+func (ddb *DoltDB) GetWorkingSets(ctx context.Context) ([]ref.DoltRef, error) {
+	return ddb.GetRefsOfType(ctx, workingSetsRefFilter)
+}
+
+// GetHeadRefs returns a list of all refs that point to a Commit
+func (ddb *DoltDB) GetHeadRefs(ctx context.Context) ([]ref.DoltRef, error) {
+	return ddb.GetRefsOfType(ctx, ref.HeadRefTypes)
 }
 
 func (ddb *DoltDB) GetRefsOfType(ctx context.Context, refTypeFilter map[ref.RefType]struct{}) ([]ref.DoltRef, error) {
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/doltdb/doltdb.go:768]: Fixed the dropped error, moved working sets out of refs/

File: go/libraries/doltcore/sqle/transactions.go (hunk 2 above)
Comment (zachmu) [reply go/libraries/doltcore/sqle/transactions.go:89]: We were not, fixed

File: go/libraries/doltcore/sqle/tablecache.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	"sync"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+)
+
+// TableCache is a caches for sql.Tables.
+// Caching schema fetches is a meaningful perf win.
+type TableCache interface {
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/sqle/tablecache.go:27]: SGTM

**Comment by zachmu** [comment]
#benchmark

**Comment by zachmu** [comment]
#benchmark

**Comment by zachmu** [comment]
#benchmark

**Comment by zachmu** [comment]
#benchmark

**Pull request merged**